GenderPayGap.WebUI/Controllers/Account/PasswordResetController.cs

```diff
@@ -0,0 +1,194 @@
+
+        [HttpGet("choose-new-password")]
+        public IActionResult ChooseNewPasswordGet(string code)
+        {
+            // Don't use the User returned from this, but call it to check that the code is valid
+            // And also has not expired
+            ExtractUserFromResetCode(code);
+
+            var viewModel = new ChooseNewPasswordViewModel {ResetCode = code};
+
+            return View("ChooseNewPassword", viewModel);
+        }
+        
+        [HttpGet("choose-new-password/complete")]
+        public IActionResult ChooseNewPasswordCompleteGet()
+        {
+            return View("ChooseNewPasswordComplete");
+        }
+        
+        [ValidateAntiForgeryToken]
+        [PreventDuplicatePost]
+        [HttpPost("choose-new-password")]
+        public IActionResult ChooseNewPasswordPost(ChooseNewPasswordViewModel viewModel)
+        {
+            viewModel.ParseAndValidateParameters(Request, m => m.NewPassword); 
+            viewModel.ParseAndValidateParameters(Request, m => m.ConfirmNewPassword);
+            
+            if (viewModel.HasSuccessfullyParsedValueFor(m => m.NewPassword)
+                && viewModel.HasSuccessfullyParsedValueFor(m => m.ConfirmNewPassword)
+                && viewModel.NewPassword != viewModel.ConfirmNewPassword)
+            {
+                viewModel.AddErrorFor(m => m.ConfirmNewPassword, "Password and confirmation password do not match");
+            }
+
+            if (viewModel.HasAnyErrors())
+            {
+                return View("ChooseNewPassword", viewModel);
+            }
+            
+            // Find the user from the reset code in the viewModel
+            User userToUpdate = ExtractUserFromResetCode(viewModel.ResetCode);
+            
+            userRepository.UpdatePassword(userToUpdate, viewModel.NewPassword);
+            emailSendingService.SendResetPasswordCompletedEmail(userToUpdate.EmailAddress);
+
+            return RedirectToAction("ChooseNewPasswordCompleteGet");
+        }
+        
+        // Look up the reset code (a GUID) in the database, and return the user it's associated with
+        private User ExtractUserFromResetCode(string encryptedCode)
+        {
+            User user = dataRepository.GetAll<User>().FirstOrDefault(u => u.PasswordResetCode == encryptedCode);
+            
+            // Check that user exists
+            if (user.IsNull())
+            {
+                throw new PageNotFoundException();
+            }
+
+            // Check that password reset code has not expired
+            if (!PasswordResetCodeHasExpired(user))
+            {
+                throw new PasswordResetCodeExpiredException();
+            }
+
+            // Remove password reset code and send date
+            user.PasswordResetCode = null;
+            user.ResetSendDate = null;
+
+            return user;
+        }
+        
+        private static bool PasswordResetEmailSentRecently(User userForPasswordReset)

```

Comment:
I might call this `PasswordResetEmailSentTooRecently` - to make it clear that it's a problem.
Also, it could go further up in the file, nearer to where it's used - then there's 2 clear sections (obtaining a reset code, using the reset code)